db/fdb_fend.c

```diff
@@ -5140,3 +5288,489 @@ void fdb_systable_info_free(void *data, int npoints)
+static int _fdb_cdb2api_send_set(fdb_cursor_t *fdbc)
+{
+    cdb2_hndl_tp *hndl = fdbc->fcon.api.hndl;
+    int rc = FDB_NOERR;
+
+    /* table version check */
+    char str[256];
+    snprintf(str, sizeof(str), "SET REMSQL_TABLE %s %d",
+             fdbc->ent ? fdbc->ent->tbl->name : "sqlite_master",
+             fdbc->ent ? fdb_table_version(fdbc->ent->tbl->version) : 0);
+
+    rc = cdb2_run_statement(hndl, str);
+    if (rc) {
+        logmsg(LOGMSG_ERROR, "%s failed to set remote table rc %d\n",
+               __func__, rc);
+        return FDB_ERR_GENERIC;
+    }
+
+    uuidstr_t us;
+    comdb2uuidstr(fdbc->ciduuid, us);
+    snprintf(str, sizeof(str), "SET REMSQL_CURSOR %s", us);
+
+    rc = cdb2_run_statement(hndl, str);
+    if (rc) {
+        logmsg(LOGMSG_ERROR, "%s failed to set cursor id rc %d\n",
+               __func__, rc);
+        return FDB_ERR_GENERIC;
+    }
+
+    return FDB_NOERR;
+}
+
+#define SET_INT(name, value) \
+    do { \
+        snprintf(str, sizeof(str), "SET %s %d", \
+                 name, value); \
+        rc = cdb2_run_statement(hndl, str); \
+        if (rc) { \
+            logmsg(LOGMSG_ERROR, "%s failed to write %s \"%s\"\n", \
+                   __func__, name, str); \
+            return -1; \
+        } \
+    } while (0);
+
+#define SET_STR(name, value) \
+    do { \
+        snprintf(str, sizeof(str), "SET %s %s", \
+                 name, value); \
+        rc = cdb2_run_statement(hndl, str); \
+        if (rc) { \
+            logmsg(LOGMSG_ERROR, "%s failed to write %s \"%s\"\n", \
+                   __func__, name, str); \
+            return -1; \
+        } \
+    } while (0);
+
+static int _fdb_client_set_options(struct sqlclntstate *clnt,
+                                   cdb2_hndl_tp *hndl)
+{
+    char str[256];
+    int rc = 0;
+
+    /* we only pass a subset of SET options */
+    if (clnt->query_timeout) {
+        SET_INT("MAXQUERYTIME", clnt->query_timeout);
+    }
+    if (clnt->tzname[0]) {
+        SET_STR("TIMEZONE", clnt->tzname);
+    }
+    char *dtprec = clnt->dtprec == DTTZ_PREC_MSEC ? "M" : "U";
+    SET_STR("DATETIME PRECISION", dtprec);
+    if (clnt->prepare_only) {
+        SET_STR("PREPARE_ONLY", "ON");
+    }
+    return 0;
+}
+
+const char *err_precdb2api = "Invalid set command 'REMSQL";
+const char *err_cdb2apiold = "need protocol ";
+const char *err_tableschemaold = "need table schema ";
+
+static int _fdb_run_sql(BtCursor *pCur, char *sql)
+{
+    fdb_cursor_t *fdbc = pCur->fdbc->impl;
+    cdb2_hndl_tp *hndl = fdbc->fcon.api.hndl;
+    sqlclntstate_fdb_t *state = pCur->clnt ? &pCur->clnt->fdb_state : NULL;
+    int rc;
+    const char *errstr;
+    unsigned long long start_rpc;
+    unsigned long long end_rpc;
+    const char *tmp;
+
+    start_rpc = osql_log_time();
+
+    /* remsql set otions */
+    rc = _fdb_cdb2api_send_set(fdbc);
+    if (rc)
+        return rc;
+
+    /* client set options */
+    rc = _fdb_client_set_options(pCur->clnt, hndl);
+
```

Comment:
we can add this logic before cdb2_run_statement to make it work over IAM enabled databases
https://github.com/bloomberg/comdb2/blob/f79e2c54608b693bf67c1387cc583cc2f7c76ccf/db/fdb_push.c#L287

Reply:
done, thank you